Adding numpy loader changes

SundarRajan28 opened the pull request.

SundarRajan28 committed:
Adding numpy loader changes

fiona-gladwin commented on rocAL/include/loaders/image/node_numpy_loader.h:
```diff
@@ -0,0 +1,57 @@
+/*
+Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
```
Copyright change to 2025 in all files

fiona-gladwin commented on rocAL/include/loaders/image/node_numpy_loader.h:
```diff
@@ -0,0 +1,57 @@
+
+#pragma once
+#include "numpy_loader_sharded.h"
+#include "pipeline/graph.h"
```
graph.h is already included in node.h

fiona-gladwin commented on rocAL/include/loaders/image/node_numpy_loader.h:
```diff
@@ -0,0 +1,57 @@
+
+class NumpyLoaderNode : public Node {
+   public:
+    /// \param device_resources shard count from user
```
Please check the description, seems off

fiona-gladwin commented on rocAL/include/loaders/image/node_numpy_loader.h:
```diff
@@ -0,0 +1,57 @@
+    std::shared_ptr<LoaderModule> get_loader_module();
+
+   protected:
+    void create_node() override {};
```
Remove semicolon at the end for these 2 lines

fiona-gladwin commented on rocAL/include/loaders/image/node_numpy_loader_single_shard.h:
Refer to same comments as numpy_loader.h

fiona-gladwin commented on rocAL/include/loaders/image/numpy_loader.h:
```diff
@@ -0,0 +1,86 @@
+#include "loaders/circular_buffer.h"
+#include "pipeline/commons.h"
+
+// NumpyLoader runs an internal thread for loading an decoding of numpy arrays asynchronously
```
loading and decoding

fiona-gladwin commented on rocAL/include/readers/image/image_reader.h:
```diff
@@ -106,14 +108,18 @@ struct ReaderConfig {
     void set_sharding_info(const ShardingInfo& sharding_info) {
         _sharding_info = sharding_info;
     }
+    void set_files(const std::vector<std::string> &files) { _files = files; }
```
Can we call it set_files_list as we are storing a list of file path here

fiona-gladwin commented on rocAL/include/readers/image/image_reader.h:
```diff
@@ -166,6 +174,26 @@ struct ImageRecordIOHeader {
+    bool fortran_order = false;
+    int64_t data_offset = 0;
+
+    RocalTensorDataType type() const { return type_info; };
```
Remove semicolon at end

fiona-gladwin commented on rocAL/include/readers/image/image_reader.h:
```diff
@@ -166,6 +174,26 @@ struct ImageRecordIOHeader {
+    RocalTensorDataType type() const { return type_info; };
+
+    size_t size() const {
+        size_t num_elements = 1;
```
Is it the size we are calculating or num_elements?
Variable name seems confusing

SundarRajan28 replied:
nbytes() function calculates the byte size for the tensor
size() function calculates the n_elements in the tensor

SundarRajan28 committed:
Change copyright year

SundarRajan28 committed:
Merge remote-tracking branch 'origin/develop' into numpy_loader_changes

SundarRajan28 committed:
Resolving review comments

shobana-mcw commented on rocAL/include/readers/image/image_reader.h:
```diff
@@ -49,6 +50,7 @@ enum class StorageType {
     VIDEO_FILE_SYSTEM = 8,
     EXTERNAL_FILE_SOURCE = 9,      // to support reading from external source
     WEBDATASET_RECORDS = 10, // tar files - webdataset format
+    NUMPY_DATA = 11
```
Add a comment here.

shobana-mcw commented on rocAL/include/readers/image/image_reader.h:
```diff
@@ -148,6 +154,8 @@ struct ReaderConfig {
     std::shared_ptr<MetaDataReader> _meta_data_reader = nullptr;
     ExternalSourceFileMode _file_mode = ExternalSourceFileMode::NONE;
     ShardingInfo _sharding_info;
+    std::vector<std::string> _files;
```
Can we rename this to file_names?

shobana-mcw commented on rocAL/include/readers/image/image_reader.h:
```diff
@@ -166,6 +174,26 @@ struct ImageRecordIOHeader {
+        return num_elements;
+    };
+
+    size_t nbytes() const { return tensor_data_size(type_info) * size(); }
```
Follow the same naming. Please change this to num_bytes similar to num_elements.

shobana-mcw commented on rocAL/include/readers/image/image_reader.h:
```diff
@@ -243,6 +271,10 @@ class Reader {
     //! Copies the data of the opened item to the buf
     virtual size_t read_data(unsigned char *buf, size_t read_size) = 0;
 
+    virtual const NumpyHeaderData get_numpy_header_data() { return {}; }
```
Add comments for these functions.

shobana-mcw commented on rocAL/include/readers/image/image_reader.h:
```diff
@@ -166,6 +174,26 @@ struct ImageRecordIOHeader {
 
+struct NumpyHeaderData {
+   public:
+    std::vector<unsigned> array_shape;
```
Add comments for member functions and variables.

shobana-mcw reviewed (commented):
Please address the review comments.

SundarRajan28 committed:
Resolve review comments

SundarRajan28 committed:
Merge branch 'develop' into numpy_loader_changes

fiona-gladwin committed:
Video - rocDecode integration in rocAL (#253)

* Add CMakeList changes

Requires VideoDemuxer to have .cpp file

* Add rocDecode utils files and cmake changes

* rocDecode integration for rocAL

* merge with tot

* Add initial support for rocDecode integration

* Add fix to decode sequences

* VideoDemuxer - support to return dts

* Support to seek exact frame

Facing seg fault

* Add intermediary buffers to copy

* Change the name of rocdec files

* Temporary rocDecode cmake changes

Add utils folder to rocAL hip to compile with rocDecode

* Fix CMake for rocDecode

* Minor fix

* Fix stride for rocdecode video decoder

* Revert "Temporary rocDecode cmake changes"

This reverts commit 21b0538d855442d88708e6ca3c4e1acb13b94f98.

* Introduce CMake for rocDecode

Revert additional changes

* Fix issues with CMake

* rocDecode Flush decoder after decoding each sequence

* Add changes to flush decoder after seek

* Add support in rocDecode to seek to exact frame using dts

* Add changes in API to pass device ID to video decoder

* Add stream synchronize

* Minor changes

* Minor change

* Change rocDecode to all caps in CMake

* Minor changes

* Minor CMake change

* Minor changes

* Introduce enable rocdecode flag in CMake

* Minor fix

* Minor changes

* Modify CMake to use only ROCM_PATH

* Use Pascal case for fn names in Video Decoder and other derived classes

* Fix CMakeLists

* Minor change

* Fix rocAL_hip CMakeLists for rocDecode

* Add decoder type argument to Video readers

* Fix video metadata reader

* Minor changes

* Move default constructor to class

* Minor fix

* Update CHANGELOG and docs

* Add link to rocDecode

* Update README with rocDecode installation instructions

* Update rocAL version

* Update README with rocDecode manual installation

---------

Co-authored-by: rrawther <Rajy.MeeyakhanRawther@amd.com>
Co-authored-by: Sundar Rajan Vaithiyanathan <99159823+SundarRajan28@users.noreply.github.com>
Co-authored-by: Kiriti Gowda <kiritigowda@gmail.com>

kiritigowda committed:
Merge branch 'develop' into numpy_loader_changes

kiritigowda committed:
rocDecode - integration updates (#281)

* rocDecode - Updates

* FindrocDecode - updates

* rocDecode - compile updates

* Package - rocDecode deps added

* Setup - cleanup

* Fix Typo

* CMakeList - Fix

* Update CHANGELOG.md

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

---------

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

kiritigowda committed:
Merge branch 'develop' into numpy_loader_changes

SundarRajan28 committed:
Removing decode time from numpy loader
